### PR TITLE
z_room (1 non-matching)

### PR DESCRIPTION
**Pull request by engineer124**
This decompiles and documents the remaining room functions

### CODE REVIEW

File: src/code/z_room.c [hunk 1]
```diff
@@ -1,94 +1,579 @@
...
+
+        // Pick and sort entries by depth
+        for (i = 0; i < roomShape->numEntries; i++, roomShapeCullableEntry++) {
+
+            // Project the entry position, to get the depth it is at.
+            pos.x = roomShapeCullableEntry->boundsSphereCenter.x;
+            pos.y = roomShapeCullableEntry->boundsSphereCenter.y;
+            pos.z = roomShapeCullableEntry->boundsSphereCenter.z;
+            SkinMatrix_Vec3fMtxFMultXYZ(&play->viewProjectionMtxF, &pos, &projectedPos);
+            // SkinMatrix_Vec3fMtxFMultXYZW(&play->viewProjectionMtxF, &pos, &projectedPos, &projectedW);
```
Comment (AngheloAlf) [inline src/code/z_room.c:173]: Why is this commented out?

File: src/code/z_room.c [hunk 2]
```diff
@@ -1,94 +1,579 @@
...
+                        }
+                    }
+
+                    insert++;
+                }
+            }
+        }
+
+        // if this is real then I might not be
+        R_ROOM_CULL_NUM_ENTRIES = roomShape->numEntries & 0xFFFF & 0xFFFF & 0xFFFF;
```
Comment (AngheloAlf) [inline src/code/z_room.c:234]: rofl

File: src/code/z_room.c [hunk 3]
```diff
@@ -1,94 +1,579 @@
...
+RoomShapeImageMultiBgEntry* Room_GetImageMultiBgEntry(RoomShapeImageMulti* roomShapeImageMulti, PlayState* play) {
+    Camera* activeCam = GET_ACTIVE_CAM(play);
+    s32 bgCamIndex = activeCam->bgCamDataId;
+    s16 overrideBgCamIndex;
+    Player* player;
+    RoomShapeImageMultiBgEntry* bgEntry;
+    s32 i;
+
+    overrideBgCamIndex = ((BgCamFuncData*)BgCheck_GetBgCamFuncDataImpl(&play->colCtx, bgCamIndex, BGCHECK_SCENE))
+                             ->roomImageOverrideBgCamIndex;
```
Comment (AngheloAlf) [inline src/code/z_room.c:384]: This cast is a bit fishy, what's going on?

File: src/code/z_room.c (hunk 1 above)
Comment (engineer124) [reply src/code/z_room.c:173]: Ah, this was here for reference because OoT's equivalent function uses `SkinMatrix_Vec3fMtxFMultXYZW`, I'll remove it

File: src/code/z_room.c (hunk 2 above)
Comment (engineer124) [reply src/code/z_room.c:234]: copy-pasta from OoT, I guess I can remove it. Interestingly seems to be needed in MM as well (or at least this fake trick also works in MM)

File: src/code/z_room.c (hunk 3 above)
Comment (engineer124) [reply src/code/z_room.c:384]: There was a long discussion related to both OoT and MM on whether the `BgCamData` is suppose to be `Vec3s`, `s16`, or follow the `BgCamFuncData` struct. Basically, it's just a group of `s16` numbers that are treated differently depending on where it's used. In OoT, its' used as `Vec3s` for crawlspaces, and as `BgCamFuncData` in most other places. In MM, the Vec3s part for crawlspaces is dropped.

So basically, `BgCheck_GetBgCamFuncDataImpl` gets the pointer to the raw s16 data, and then the intention is to cast it to its appropriate type. But I don't think this issue is 100% solved, and we're settled on a more intermediate solution, for now

File: include/z64scene.h [hunk 4]
```diff
@@ -213,58 +216,132 @@ typedef struct {
...
+    /* 3 */ ROOM_SHAPE_TYPE_NONE,
+    /* 4 */ ROOM_SHAPE_TYPE_MAX
+} RoomShapeType;
 
-// Size TODO
 typedef struct {
-    /* 0x0 */ UNK_TYPE1 pad0[0x10];
-} RoomMeshType1Params; // size = 0x10
+    /* 0x00 */ u8 type;
+} RoomShapeBase; // size = 0x01
```
Comment (hensldm) [inline include/z64scene.h:229]:
```suggestion
    /* 0x0 */ u8 type;
} RoomShapeBase; // size = 0x1
```

File: include/z64scene.h [hunk 5]
```diff
@@ -213,58 +216,132 @@ typedef struct {
...
+typedef enum {
+    /* 1 */ ROOM_SHAPE_IMAGE_AMOUNT_SINGLE = 1,
+    /* 2 */ ROOM_SHAPE_IMAGE_AMOUNT_MULTI
+} RoomShapeImageAmountType;
+
+typedef struct {
+    /* 0x00 */ RoomShapeBase base;
+    /* 0x01 */ u8 amountType; // RoomShapeImageAmountType
+    /* 0x04 */ RoomShapeDListsEntry* entry;
+} RoomShapeImageBase; // size = 0x08
```
Comment (hensldm) [inline include/z64scene.h:252]:
```suggestion
typedef struct {
    /* 0x0 */ RoomShapeBase base;
    /* 0x1 */ u8 amountType; // RoomShapeImageAmountType
    /* 0x4 */ RoomShapeDListsEntry* entry;
} RoomShapeImageBase; // size = 0x8
```

File: src/code/z_play.c [hunk 6]
```diff
@@ -430,7 +430,7 @@ void func_80169FDC(GameState* thisx) {
 s32 func_80169FFC(GameState* thisx) {
     PlayState* this = (PlayState*)thisx;
 
-    return this->roomCtx.curRoom.mesh->type0.type != 1;
+    return this->roomCtx.curRoom.roomShape->normal.base.type != 1;
```
Comment (hensldm) [inline src/code/z_play.c:433]: Not sure, should this use the `RoomShapeType` enum?

File: include/functions.h [hunk 7]
```diff
@@ -1994,7 +1989,7 @@ void Scene_HeaderCmdMiniMapCompassInfo(PlayState* play, SceneCmd* cmd);
 void Scene_HeaderCmdSetRegionVisitedFlag(PlayState* play, SceneCmd* cmd);
 void Scene_HeaderCmdAnimatedMaterials(PlayState* play, SceneCmd* cmd);
 void Scene_SetExitFade(PlayState* play);
-s32 Scene_ProcessHeader(PlayState* play, SceneCmd* header);
+s32 Scene_ExecuteCommands(PlayState* play, SceneCmd* header);
```
Comment (hensldm) [inline include/functions.h:1992]: Should the arg name change as well?

**Review by AngheloAlf — APPROVED** [review]

File: src/code/z_play.c (hunk 6 above)
Comment (hensldm) [reply src/code/z_play.c]: Nah, that is fine just do it here :)

File: src/code/z_scene.c [hunk 8]
```diff
@@ -535,57 +535,58 @@ void Scene_SetExitFade(PlayState* play) {
     play->transitionType = Entrance_GetTransitionFlags(play->nextEntrance) & 0x7F;
 }
 
+static void (*sceneCmdHandlers[])(PlayState*, SceneCmd*) = {
+    Scene_HeaderCmdSpawnList,
```
Comment (hensldm) [inline src/code/z_scene.c:539]: Could probably rename these with the header now as well.

**Review by hensldm — APPROVED** [review]
1 more comment, but LGTM

File: src/code/z_scene.c (hunk 8 above)
Comment (engineer124) [reply src/code/z_scene.c]: Took to the OoT style and went `Scene_HeaderCmd` -> `Scene_Command`
also
`sceneCmdHandlers` -> `sSceneCmdHandlers`

Comment (fig02) [reply src/code/z_scene.c]: just for clarity, I dislike this naming scheme and have proposed `CutsceneCmd_` as a new system name in z_demo docs, and think if that is agreed on, it can be used for other command handler functions as well in other scripting systems

Not saying to use this now, but just saying that I dont think this style in oot is 100% solidified

Comment (engineer124) [reply src/code/z_scene.c]: Good to note, thanks. My preference is to just leave it as I've changed it for now, and just sync later once new names are decided on in OoT if we like them

File: include/z64scene.h [hunk 9]
```diff
@@ -213,58 +216,132 @@ typedef struct {
...
-    /* 0x8 */ RoomMeshType2Params* paramsEnd;
-} RoomMeshType2; // size = 0xC
+typedef enum {
+    /* 1 */ ROOM_SHAPE_IMAGE_AMOUNT_SINGLE = 1,
+    /* 2 */ ROOM_SHAPE_IMAGE_AMOUNT_MULTI
+} RoomShapeImageAmountType;
+
+typedef struct {
+    /* 0x0 */ RoomShapeBase base;
+    /* 0x1 */ u8 amountType; // RoomShapeImageAmountType
```
Comment (EllipticEllipsis) [inline include/z64scene.h:250]: I don't think I would have realised this was referring to the enum if it were not right next-door, something like `uses X enum` seems better.

File: include/z64scene.h [hunk 10]
```diff
@@ -213,58 +216,132 @@ typedef struct {
...
+    /* 0x16 */ u16 height;
+    /* 0x18 */ u8 fmt;
+    /* 0x19 */ u8 siz;
+    /* 0x1A */ u16 tlutMode;
+    /* 0x1C */ u16 tlutCount;
+} RoomShapeImageSingle; // size = 0x20
+
+typedef struct {
+    /* 0x00 */ u16 unk_00;
+    /* 0x02 */ u8 bgCamIndex; // for which bg cam index is this entry for
```
Comment (EllipticEllipsis) [inline include/z64scene.h:269]:
```suggestion
    /* 0x02 */ u8 bgCamIndex; // the bg cam index this entry is for
```
? Or have I misinterpreted the meaning?

File: include/z64scene.h [hunk 11]
```diff
@@ -213,58 +216,132 @@ typedef struct {
...
+
+typedef struct {
+    /* 0x00 */ RoomShapeImageBase base;
+    /* 0x08 */ void* source;
+    /* 0x0C */ u32 unk_0C;
+    /* 0x10 */ void* tlut;
+    /* 0x14 */ u16 width;
+    /* 0x16 */ u16 height;
+    /* 0x18 */ u8 fmt;
+    /* 0x19 */ u8 siz;
```
Comment (EllipticEllipsis) [inline include/z64scene.h:262]: Presumably there's a good reason these are abbreviations... named after gbi stuff?

File: include/z64scene.h [hunk 12]
```diff
@@ -290,11 +367,11 @@ typedef struct {
     struct {
         s8 room;    // Room to switch to
         s8 bgCamIndex; // How the camera reacts during the transition. -2 for spiral staircase. -1 for generic door. 0+ will index scene CamData
-    } /* 0x0 */ sides[2]; // 0 = front, 1 = back
-    /* 0x4 */ s16   id;
-    /* 0x6 */ Vec3s pos;
-    /* 0xC */ s16   rotY;
-    /* 0xE */ u16   params;
+    } /* 0x00 */ sides[2]; // 0 = front, 1 = back
```
Comment (EllipticEllipsis) [inline include/z64scene.h:370]: Ideally this gets a typedef at some point. Also didn't we decide to remove the leading 0s when the numbers didn't require them?

File: src/code/z_room.c [hunk 13]
```diff
@@ -1,94 +1,578 @@
 #include "global.h"
 
-void Room_nop8012D510(PlayState* play, Room* room, UNK_PTR param_3, UNK_TYPE1 param_4) {
+void Room_Update(PlayState* play, Room* room, Input* input, s32 arg3) {
 }
 
-void Room_DrawType3Mesh(PlayState* play, Room* room, u32 flags) {
+void Room_DrawNone(PlayState* play, Room* room, u32 flags) {
 }
```
Comment (EllipticEllipsis) [inline src/code/z_room.c:7]: I assume these are based on usage?

File: src/code/z_room.c [hunk 14]
```diff
@@ -1,94 +1,578 @@
...
+        if (flags & ROOM_DRAW_XLU) {
+            for (; head != NULL; head = head->prev) {
+                f32 temp_fv0;
+                f32 temp_fv1;
+
+                roomShapeCullableEntry = head->entry;
+                displayList = roomShapeCullableEntry->xlu;
+
+                if (displayList != NULL) {
+                    if (roomShapeCullableEntry->boundsSphereRadius & 1) {
```
Comment (EllipticEllipsis) [inline src/code/z_room.c:276]: What a weird check.

**Review by EllipticEllipsis — COMMENTED** [review]
Few minor things, most of the docs look good, I remember looking at the OoT version.

File: include/z64scene.h (hunk 12 above)
Comment (engineer124) [reply include/z64scene.h:370]: What did we decide for  `// size = 0x10`, I thought that since the total size was 2 digits, that all members should also be 2 digits? Although a quick peak though the repo and I'm seeing a lot of both cases (2 digits and 1 digits when ` // size = 0x10`?

And yeah, would be good to enum this but I'd like to wait until transition/doors/etc.. are better understood

File: src/code/z_room.c (hunk 13 above)
Comment (engineer124) [reply src/code/z_room.c:7]: Yeah, they're both inferred. `Room_Update` because it's called from `Play_Update`, there's no other update function, and presumably you'd want to put anything you want to update in this function. Just an assumption though, The arg for `Input` is weird as well. Could always revert to just a `Noop` if we want.

`Room_DrawNone` should at least follow the functions in the draw handlers, so could also do `Room_DrawNoop` here

File: include/z64scene.h (hunk 11 above)
Comment (engineer124) [reply include/z64scene.h:262]: This was taken from OoT, and given `RoomShapeImageSingle` goes entirely unused in MM, this in a sense is dependent on OoT for docs. So that'll be a @Dragorn421 question since they did these docs for OoT. But I'm also assuming it's named after gbi stuff

File: include/z64scene.h (hunk 12 above)
Comment (EllipticEllipsis) [reply include/z64scene.h:370]: I think we decided to pad to the number of digits required within the struct itself rather than including the size in that.

File: include/z64scene.h (hunk 11 above)
Comment (fig02) [reply include/z64scene.h:262]: these abbreviations are how they are commonly named in gbi macro args yes. for example
`#define gDPLoadTextureBlock(pkt, timg, fmt, siz, width, height, pal, cms, cmt, masks, maskt, shifts, shiftt)`

File: src/code/z_room.c (hunk 13 above)
Comment (fig02) [reply src/code/z_room.c:7]: the fact that it takes a controller input arg lends itself to being something related to debugging that was ifdefed out. imo not enough info to say

Comment (fig02) [reply src/code/z_room.c:7]: I would say `Room_HandleLoadCallbacks` is the closet thing to a room update (and ive already voiced my concerns about this name in the past, theres no callbacks really happening here)

Comment (engineer124) [reply src/code/z_room.c:7]: That's fair about the debug interpretation of `Input`, I can change to just `Room_Noop`, it doesn't really change all that much since it's just speculation.

I'm also definitely not attached to `Room_HandleLoadCallbacks`, but I was going to wait to change this name later in hopes of an OoT discussion to see what kind of names come up, particularly by those who work with Fast64 and really understand scenes/rooms quite well. However, if there's a good name we can come up with now, we can change it.

Comment (fig02) [reply src/code/z_room.c:7]: it can for sure wait for oot. just mentioning that function (which is also called by Play_Update) is moreso the update for rooms (even though in this context, updating a room just means doing a few things on new room load)

File: include/z64scene.h (hunk 12 above)
Comment (EllipticEllipsis) [reply include/z64scene.h]: Also typedef-wise I meant the struct, we basically never use untyped structs.

Comment (engineer124) [reply include/z64scene.h]: Oh, the internal struct? Most internal structs within structs don't seem to use a typedef, unless you wanted to move this struct externally and break this struct into two

File: src/code/z_scene.c [hunk 15]
```diff
@@ -536,57 +536,58 @@ void Scene_SetExitFade(PlayState* play) {
     play->transitionType = Entrance_GetTransitionFlags(play->nextEntrance) & 0x7F;
 }
 
+static void (*sSceneCmdHandlers[])(PlayState*, SceneCmd*) = {
```
Comment (Dragorn421) [inline src/code/z_scene.c:539]: remove static ? (idk how you handle static in MM, I've also seen it removed from overlay symbols)

File: src/code/z_room.c [hunk 16]
```diff
@@ -1,94 +1,578 @@
...
+        if (play->roomCtx.unk74 != NULL) {
+            gSPSegment(POLY_XLU_DISP++, 0x06, play->roomCtx.unk74);
+        }
+        func_8012C2DC(play->state.gfxCtx);
+        gSPMatrix(POLY_XLU_DISP++, &gIdentityMtx, G_MTX_MODELVIEW | G_MTX_LOAD);
+    }
+
+    if ((room->enablePosLights != 0) || (MREG(93) != 0)) {
+        gSPSetGeometryMode(POLY_OPA_DISP++, 0x00400000);
+        gSPSetGeometryMode(POLY_XLU_DISP++, 0x00400000);
```
Comment (Dragorn421) [inline src/code/z_room.c:116]:
```suggestion
        gSPSetGeometryMode(POLY_OPA_DISP++, G_LIGHTING_POSITIONAL);
        gSPSetGeometryMode(POLY_XLU_DISP++, G_LIGHTING_POSITIONAL);
```

File: src/code/z_room.c [hunk 17]
```diff
@@ -1,94 +1,578 @@
...
+    size_t backRoomSize;
+    size_t cumulRoomSize;
+    s32 pad[2];
+
+    {
+        RomFile* roomList = play->roomList;
+
+        for (i = 0; i < play->numRooms; i++) {
+            roomSize = roomList[i].vromEnd - roomList[i].vromStart;
+            maxRoomSize = CLAMP_MIN(maxRoomSize, roomSize);
```
Comment (Dragorn421) [inline src/code/z_room.c:513]: does `MAX` not match here? would make more sense when reading imo

same below

File: src/code/z_room.c [hunk 18]
```diff
@@ -1,94 +1,578 @@
...
+        }
+    }
+
+    if (play->doorCtx.numTransitionActors) {
```
Comment (Dragorn421) [inline src/code/z_room.c:517]: does it need to not have `!= 0`

File: src/code/z_room.c [hunk 19]
```diff
@@ -1,94 +1,578 @@
...
+            maxRoomSize = CLAMP_MIN(maxRoomSize, cumulRoomSize);
+            transitionActor++;
+        }
+    }
+
+    roomCtx->roomMemPages[0] = THA_AllocEndAlign16(&play->state.heap, maxRoomSize);
+    if (roomCtx->roomMemPages[0] == NULL) {
+        __assert("../z_room.c", 1078);
+    }
+    roomCtx->roomMemPages[1] = (void*)((s32)roomCtx->roomMemPages[0] + maxRoomSize);
```
Comment (Dragorn421) [inline src/code/z_room.c:537]: Since you used size_t I guess you'd want to use (u)intptr_t here

**Review by Dragorn421 — COMMENTED** [review]
I didn't know MM had `RoomContext.roomMemPages` (OoT's `RoomContext.bufPtrs`), interesting name (though "page" here is not https://en.wikipedia.org/wiki/Page_%28computer_memory%29 )

File: src/code/z_scene.c (hunk 15 above)
Comment (engineer124) [reply src/code/z_scene.c:539]: I don't think it's set and stone once data is named, but I think the tendency is to not have it if it's a unique enough name, but it's not been a huge deal. This is unique enough to remove though, so I'll do it

File: src/code/z_room.c (hunk 16 above)
Comment (engineer124) [reply src/code/z_room.c:116]: Huh, not sure why that was missed

File: src/code/z_room.c (hunk 18 above)
Comment (engineer124) [reply src/code/z_room.c:517]: It needs it without the `!= 0` to match, or the `(u32)` trick still being discussed in OoT

File: include/z64scene.h (hunk 12 above)
Comment (AngheloAlf) [reply include/z64scene.h]: I think we have a few unnamed structs inside a few named structs, like a few in audio, bgCheck, camera, interface, effect, etc.
You can find them with this regex: `[^f] struct \{`

**Review by Dragorn421 — APPROVED** [review]

File: src/code/z_scene.c [hunk 20]
```diff
@@ -168,19 +168,19 @@ void Scene_HeaderCmdSpawnList(PlayState* play, SceneCmd* cmd) {
...
 
 // SceneTableEntry Header Command 0x02: List of camera data for actor cutscenes
-void Scene_HeaderCmdActorCutsceneCamList(PlayState* play, SceneCmd* cmd) {
+void Scene_CommandActorCutsceneCamList(PlayState* play, SceneCmd* cmd) {
     play->actorCsCamList = Lib_SegmentedToVirtual(cmd->actorCsCamList.segment);
 }
 
 // SceneTableEntry Header Command 0x03: Collision Header
-void Scene_HeaderCmdColHeader(PlayState* play, SceneCmd* cmd) {
+void Scene_CommandColHeader(PlayState* play, SceneCmd* cmd) {
```
Comment (mzxrules) [inline src/code/z_scene.c:183]:
```suggestion
void Scene_CommandCollisionHeader(PlayState* play, SceneCmd* cmd) {
```

File: src/code/z_play.c [hunk 21]
```diff
@@ -2302,7 +2301,8 @@ void Play_Init(GameState* thisx) {
 
     while (!Room_HandleLoadCallbacks(this, &this->roomCtx)) {}
 
-    if ((CURRENT_DAY != 0) && ((this->roomCtx.curRoom.unk3 == 1) || (this->roomCtx.curRoom.unk3 == 5))) {
+    if ((CURRENT_DAY != 0) && ((this->roomCtx.curRoom.behaviorType1 == ROOM_BEHAVIOR_TYPE1_1) ||
+                               (this->roomCtx.curRoom.behaviorType1 == ROOM_BEHAVIOR_TYPE1_5))) {
         Actor_Spawn(&this->actorCtx, this, ACTOR_EN_TEST4, 0.0f, 0.0f, 0.0f, 0, 0, 0, 0);
```
[the remark this answers is not in the document]
Comment (hensldm) [reply src/code/z_play.c:2306]: > also the inside of the clocktower, where there is no time actor, is type 1....
KZ shows `ACTOR_EN_TEST4` is actually spawned inside the clocktower.

File: src/code/z_room.c [hunk 22]
```diff
@@ -1,94 +1,578 @@
...
+    /* 0x08 */ struct RoomShapeCullableEntryLinked* prev;
+    /* 0x0C */ struct RoomShapeCullableEntryLinked* next;
+} RoomShapeCullableEntryLinked; // size = 0x10
+
+// TODO: 127 is an arbitrarily chosen number to make the stack sorta work
+#define ROOM_SHAPE_CULLABLE_MAX_ENTRIES 127
+
+#ifdef NON_MATCHING
+// Small regalloc, likely related to temp usage and scoping
+void Room_DrawCullable(PlayState* play, Room* room, u32 flags) {
```
Comment (EllipticEllipsis) [inline src/code/z_room.c:67]: I don't suppose this yields to any do while (0) shenanigans in the macros?

**Review by EllipticEllipsis — APPROVED** [review]

File: src/code/z_room.c (hunk 22 above)
Comment (engineer124) [reply src/code/z_room.c:67]: Nothing stands out, but I can try looking at it. I don't think I looked at debug for this either, pretty sure we can get `ROOM_SHAPE_CULLABLE_MAX_ENTRIES` from it most likely, add it to the TODO
